notebooks/ntsb-duckdb.ipynb

```diff
@@ -0,0 +1,340 @@
+    "\n",
+    "ds = (\n",
+    "    ctx.read.binary(paths, binary_format=\"pdf\", filesystem=fsys)\n",
+    "    # Parition with the sycamore partitioner, pulling out tables and images. ActorPoolStrategy(size=3) works best on my particular hardware\n",
+    "    # but your mileage may vary depending on your RAM.\n",
+    "    .partition(partitioner=SycamorePartitioner(extract_table_structure=True, extract_images=True, use_cache=False), compute=ActorPoolStrategy(size=3))\n",
+    "    # Get rid of spurious whitespace charaters\n",
+    "    .regex_replace(COALESCE_WHITESPACE)\n",
+    "    # Automatically determine a schema of additional metadata to extract from documents\n",
+    "    .extract_batch_schema(schema_extractor=OpenAISchemaExtractor(\"FlightAccidentReport\", llm=llm, num_of_elements=35))\n",
+    "    # Extract the metadata specified by that schema\n",
+    "    .extract_properties(property_extractor=OpenAIPropertyExtractor(llm=llm, num_of_elements=35))\n",
+    "    # Merge elements into larger chunks\n",
+    "    .mark_bbox_preset(tokenizer=tokenizer)\n",
+    "    .merge(merger=MarkedMerger())\n",
+    "    # Convert extracted timestamps to better-structured form using the function above\n",
+    "    .map(convert_timestamp)\n",
+    "    # Copy document properties to each document's sub-elements\n",
+    "    .spread_properties([\"path\", \"entity\"])\n",
+    "    # Split elements that are too big to embed\n",
+    "    .split_elements(tokenizer=tokenizer, max_tokens=512)\n",
+    "    # Convert all Elements to Documents\n",
+    "    .explode()\n",
+    "    # Generate a series of hashes to represent each document. For use with near-duplicate detection\n",
+    "    .sketch()\n",
+    "    # Embed each document\n",
+    "    .embed(embedder=SentenceTransformerEmbedder(model_name=\"sentence-transformers/all-MiniLM-L6-v2\", batch_size=100))\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Write to a persistent duckdb database\n",
+    "# - into a specific database (as specified by url) \n",
+    "# - into a specific table (as specified by table name)\n",
+    "persistent_db = \"demo.db\"\n",
+    "persistent_table = \"demo_table\"\n",
+    "ds.write.duckdb(\n",
+    "    db_url=persistent_db,\n",
+    "    table_name=persistent_table,\n",
+    "    dimensions=384\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# We connect to the DuckDB to perform operations\n",
+    "import duckdb\n",
+    "data_conn = duckdb.connect(\"demo.db\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# DuckDB is not production ready yet for efficient Vector Similarity Search on persistent databases (https://duckdb.org/docs/extensions/vss.html), \n",
```

Comment:
Prod ready sounds weird, as per our prior convo. Would recommend: "DuckDB's efficient Vector Similarity Search on persistent databases is still an experimental feature (https://duckdb.org/docs/extensions/vss.html), so we load our data into an in-memory database. Once this feature is more stable, we can query the persistent database directly using VSS."